### PR TITLE
Exp/cache

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

File: datafusion/physical-plan/src/joins/cross_join.rs [hunk 1]
```diff
@@ -95,6 +97,41 @@ impl CrossJoinExec {
...
+            self.right.equivalence_properties().clone(),
+            &JoinType::Full,
+            self.schema(),
+            &[false, false],
+            None,
+            &[],
+        );
+
+        // Output Partitioning
+        // TODO optimize CrossJoin implementation to generate M * N partitions
```
Comment (metesynnada) [inline datafusion/physical-plan/src/joins/cross_join.rs:115]: Is this valid?

File: datafusion/physical-plan/src/joins/cross_join.rs [hunk 2]
```diff
@@ -95,6 +97,41 @@ impl CrossJoinExec {
     pub fn right(&self) -> &Arc<dyn ExecutionPlan> {
         &self.right
     }
+
+    fn with_cache(mut self) -> Self {
+        // Equivalence Properties
+        // TODO: Check equivalence properties of crossjoin, for some cases it may preserve ordering.
```
Comment (metesynnada) [inline datafusion/physical-plan/src/joins/cross_join.rs:103]: We can discuss this.

File: datafusion/physical-plan/src/values.rs [hunk 3]
```diff
@@ -109,16 +112,32 @@ impl ValuesExec {
...
         self.data.clone()
     }
+
+    fn with_cache(mut self) -> Self {
+        let mut new_cache = self.cache;
+        // Output Partitioning
+        new_cache = new_cache.with_partitioning(Partitioning::UnknownPartitioning(1));
+
+        // Execution Mode
+        let exec_mode = ExecutionMode::Bounded;
```
Comment (metesynnada) [inline datafusion/physical-plan/src/values.rs:135]: ValuesExec does not have child, so usage is correct.

File: datafusion/physical-plan/src/work_table.rs [hunk 4]
```diff
@@ -104,8 +108,22 @@ impl WorkTableExec {
...
     }
+
+    fn with_cache(mut self) -> Self {
+        let mut new_cache = self.cache;
+        // Output Partitioning
+        new_cache = new_cache.with_partitioning(Partitioning::UnknownPartitioning(1));
+
+        // Execution Mode
+        let exec_mode = ExecutionMode::Bounded;
+        new_cache = new_cache.with_exec_mode(exec_mode);
```
Comment (metesynnada) [inline datafusion/physical-plan/src/work_table.rs:122]:
```suggestion
        new_cache = new_cache.with_exec_mode(ExecutionMode::Bounded);
```

File: datafusion/sqllogictest/test_files/window.slt [hunk 5]
```diff
@@ -3580,7 +3580,7 @@ SortPreservingMergeExec: [c@3 ASC NULLS LAST]
...
 ----BoundedWindowAggExec: wdw=[AVG(multiple_ordered_table_inf.d) PARTITION BY [multiple_ordered_table_inf.d] ORDER BY [multiple_ordered_table_inf.a ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND CURRENT ROW: Ok(Field { name: "AVG(multiple_ordered_table_inf.d) PARTITION BY [multiple_ordered_table_inf.d] ORDER BY [multiple_ordered_table_inf.a ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND CURRENT ROW", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int32(10)), end_bound: CurrentRow }], mode=[Linear]
 ------CoalesceBatchesExec: target_batch_size=4096
---------RepartitionExec: partitioning=Hash([d@4], 2), input_partitions=2, preserve_order=true, sort_exprs=a@1 ASC NULLS LAST,b@2 ASC NULLS LAST
+--------RepartitionExec: partitioning=Hash([d@4], 2), input_partitions=2, preserve_order=true, sort_exprs=a@1 ASC NULLS LAST,b@2 ASC NULLS LAST,c@3 ASC NULLS LAST
```
Comment (metesynnada) [inline datafusion/sqllogictest/test_files/window.slt:3583]: How this output ordering added while it is not ordered in source?

**Review by metesynnada — COMMENTED** [review]
LGTM.

File: datafusion-cli/src/exec.rs [hunk 6]
```diff
@@ -237,7 +236,7 @@ async fn exec_and_print(
         let df = ctx.execute_logical_plan(plan).await?;
         let physical_plan = df.create_physical_plan().await?;
 
-        if is_plan_streaming(&physical_plan)? {
+        if physical_plan.unbounded_output().is_executable() {
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion-cli/src/exec.rs]: Fixed it

File: datafusion/physical-plan/src/lib.rs [hunk 7]
```diff
@@ -446,6 +450,118 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
...
 
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum ExecutionMode {
+    Bounded,
+    Unbounded,
+    InExecutable,
+}
+
+impl ExecutionMode {
+    pub fn is_unbounded(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/physical-plan/src/lib.rs:461]: Fixed it

File: datafusion/physical-plan/src/lib.rs [hunk 8]
```diff
@@ -446,6 +450,118 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
...
+    let modes = children
+        .into_iter()
+        .map(|child| child.unbounded_output())
+        .collect::<Vec<_>>();
+    exec_mode_flatten_helper(&modes)
+}
+
+#[derive(Debug, Clone)]
+pub struct PlanPropertiesCache {
+    pub eq_properties: EquivalenceProperties,
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/physical-plan/src/lib.rs:537]: Removed duplication

File: datafusion/sqllogictest/test_files/window.slt (hunk 5 above)
Comment (mustafasrepo) [reply datafusion/sqllogictest/test_files/window.slt:3583]: Source is a bit misleading. Fixed it

**Comment by mustafasrepo** [comment]
Code moved to the [PR](https://github.com/synnada-ai/datafusion-upstream/pull/1)

**Pull request closed without merging**